Update Exec and Portforward client to use pod subresource

csrwng opened the pull request:
Switches cli cmd for portfoward and exec to use /pods/[name]/exec and /pod/[name]/portforward

csrwng committed:
Update Exec and Portforward client to use pod subresource

csrwng commented:
@smarterclayton ptal

smarterclayton commented on pkg/client/remotecommand/remotecommand.go:
```diff
@@ -66,26 +68,26 @@ func New(req *client.Request, config *client.Config, command []string, stdin io.
 
 	if doStdin {
-		e.req.Param(api.ExecStdinParam, "1")
+		e.req.Param("stdin", "true")
```
Where'd these constants go?  Deleted?  If so, you should be creating an api.ExecOptions and then converting it to the desired API version, then converting it to a map.

csrwng replied:
@smarterclayton they weren't deleted, just not applicable anymore. They were for the internal endpoint. I will use the api.ExecOptions object to do the conversion.

csrwng committed:
Utility to convert versioned runtime objects to a set query parameters

JSON struct tags are used as parameter names, fields that do not have
the omitempty marker are always included.

csrwng commented:
Updated parameter conversion for exec

csrwng commented:
Moved conversion code under pkg/conversion

smarterclayton commented on pkg/kubectl/cmd/portforward.go:
```diff
@@ -43,31 +44,58 @@ $ kubectl port-forward -p mypod 0:5000`
-func RunPortForward(f *cmdutil.Factory, cmd *cobra.Command, args []string) error {
-	podName := cmdutil.GetFlagString(cmd, "pod")
-	if len(podName) == 0 {
+type portForwardCmd struct {
```
Why this abstraction?  Seems overly complex, not sure what you're trying to do here.

csrwng replied:
Only reason is to test... abstract out call to portforward.New(...) and following call to that object's ForwardPorts method. I'll think of a way to simplify it.

csrwng commented:
complexity reduced

smarterclayton commented on pkg/kubectl/cmd/portforward.go:
```diff
@@ -59,12 +60,25 @@ func NewCmdPortForward(f *cmdutil.Factory) *cobra.Command {
+	return fw.ForwardPorts()
+}
+
+func runPortForward(f *cmdutil.Factory, cmd *cobra.Command, args []string, fw portForwarder) error {
```
Why is this no longer public?

csrwng replied:
doesn't need to be. It's only used by the command.

smarterclayton replied:
Our pattern is that all of these are public.

smarterclayton commented on pkg/kubectl/cmd/exec.go:
```diff
@@ -39,36 +40,53 @@ $ kubectl exec -p 123456-7890 -c ruby-container -i -t -- bash -il`
+	tty           bool
+}
+
+func runExec(f *cmdutil.Factory, cmd *cobra.Command, cmdIn io.Reader, cmdOut, cmdErr io.Writer, p *execParams, args []string, re remoteExecutor) error {
```
Should still be RunExec

csrwng commented:
reverted to public functions

smarterclayton commented:
LGTM
